### ISSUE TITLE
TSDB Commit should honor ordering between floats and native histograms

### ISSUE THREAD

**Issue by krajorama** [prometheus/prometheus#15177]
## Background

Spun out of #15142.
Related to #15112.

The current tsdb/head_append.go headAppender.Commit processes floats first, then integer native histograms and then float native histograms. This might not be the order of the samples actually.

There is a side effect in the memSeries.lastHistogram/memSeries.lastFloatHistogram handling as well, more details in  #15142 .

## Requirements

The desired solution would:
- honor the ordering,
- set lastHistogram/lastFloatHistogram in Commit() (possible WAL replay) only,
- turn float stale markers into histograms in Commit() instead of append.

## Implementation idea

The trivial idea would be to add one more slice that specifies the type of the next sample. We'd iterate the new array and take the next sample from the appropriate slice of `a.samples` , `a.histograms` or `a.floatHistograms`.
Optimize by not storing the type for every sample, rather remember how many samples are of the next type. For example for all floats this would just say that the type is float and the number is the same as `len(a.samples)`.

**Comment by krajorama** [comment]
I think this issue causes this as well: https://github.com/prometheus/prometheus/blob/bfbae39931a6ddeb5913a6ea8d48a34bbebc6d29/storage/remote/write_handler.go#L440

**Comment by krajorama** [comment]
I've given this a try: https://github.com/prometheus/prometheus/pull/16823.  However the trivial idea above isn't enough. An appender handles multiple series, so even if the [headAppender](https://github.com/prometheus/prometheus/blob/c481aaf762bf24155d297a3efdaef5ebc61aeba0/tsdb/head_append.go#L315) keeps the order, the records that hold the data will need to be in the right order as well in the WAL. With the current different record types in the WAL, it would mean switching between record types in the WAL, which doesn't seem very optimal (haven't measured it). Probably the right solution would be to have a common [record type](https://github.com/prometheus/prometheus/blob/c481aaf762bf24155d297a3efdaef5ebc61aeba0/tsdb/record/record.go#L36) in the WAL for all samples.

**Comment by crush-on-anechka** [comment]
Am I correct that your idea is to introduce a new WAL record type, let's say `Mixed` or `MixedSamples` (`Type = 9`), and implement an encoder that tags each entry with its type, plus a decoder that can handle all types in order?

edit: read #16823 more carefully. Am I correct that “clever splitting” is more of a workaround that may introduce some overhead, while a new union record type would be a more robust long-term solution?

**Comment by krajorama** [comment]
> Am I correct that your idea is to introduce a new WAL record type, let's say `Mixed` or `MixedSamples` (`Type = 9`), and implement an encoder that tags each entry with its type, plus a decoder that can handle all types in order?
> 
> edit: read [#16823](https://github.com/prometheus/prometheus/pull/16823) more carefully. Am I correct that “clever splitting” is more of a workaround that may introduce some overhead, while a new union record type would be a more robust long-term solution?

Yes to both, as in that's the two ideas, beside the 3rd option where we don't solve this https://github.com/prometheus/prometheus/issues/14172#issuecomment-3136492345 .

The first idea with the new WAL record type would solve the issue in a straightforward way, but at the cost of disk space. I'm not sure how much. This would not be backwards compatible, so introducing it is tricky. To be able to downgrade from a version that introduces it, we'd have to either ensure upgrade path (support reading the new record type in latest LTS, but not write it, then start writing it in next LTS ) or more likely introduce via feature flag (see discussion in https://github.com/prometheus/proposals/pull/40 cc @machine424 @bwplotka ).

On the other hand the clever splitting assumes that you can detect the problem in the appender in the first place to actually split the samples into the correct record types in the correct order. The problem is that the only way to do it is to add a map of the series (plus second map for hash conflict resolution) and their state in the appender , essentially mirroring TSDB head - without the locking since appender is single thread. Then at commit you'd have to batch together samples to not flip-flop between record types needlessly, since that would introduce a lot of overhead.  This seems pretty complicated when you consider how rare this problem is. Overhead is unknown here as well.

I slightly prefer the first option just because of less complexity, disk being relatively cheap and a more straightforward adoption into remote write later down the line. cc @beorn7 @juliusmh

**Comment by beorn7** [comment]
Introducing a whole new WAL type just to handle this rarely occurring problem definitely rubs me the wrong way.

Another nuance here is that we might very well end up in a bright future where every series has a `__type__` label so that we don't have this problem in the first place (at least when it comes to mixing simple float samples and histogram samples – we still have to pay attention to not create the same problem again between integer histograms and float histograms).

> Then at commit you'd have to batch together samples to not flip-flop between record types needlessly, since that would introduce a lot of overhead.

Since the change in type is so exceedingly rare (unless there is a misconfiguration), I think we can accept a lot of overhead in the case of changing types. We only have to make sure that we don't punish the regular case of no type changes too much.

Coming back to @krajorama's comment very far above:

> With the current different record types in the WAL, it would mean switching between record types in the WAL, which doesn't seem very optimal (haven't measured it).

Given that this will happen very rarely, perhaps it doesn't matter that it is not "very optimal"? Maybe the approach in #16823 is viable after all…

**Comment by beorn7** [comment]
I have raised this to P2 because it is now the umbrella issue. (#14172 closed as a duplicate of this.)

**Comment by beorn7** [comment]
The plan for now is to do something in the spirit of #16823 in the appender, i.e. make sure that the appender uses the same write order during committing as during appending. There are different ways of doing this, but we are confident it is possible to do with acceptable overhead.

This should solve writing to TSDB, but when writing to the WAL, we run into the same issue that samples of the same type are batched together and therefore we lose the order in the rare case where the same series contains mixed types. Fundamentally, we have to start a new record whenever any given time series switches type (which is an exceedingly rare case in practice, so the overhead for starting a new record is irrelevant). To do so, we have to do some tracking during WAL writing, so we notice that case and can start a new record. We don't have a precise idea right now how to implement that, but we feel it's worth pursuing. One idea is that we could limit the size of a record to some reasonable number without a lot of overhead. Then we only have to do the tracking within that limited number, so we can prevent unbounded memory consumption for the tracking.

**Comment by beorn7** [comment]
I have a different idea how to approach this. GIving it a spin right now...

**Comment by crush-on-anechka** [comment]
Happy to help out with some of the work if that would be useful

**Comment by beorn7** [comment]
https://github.com/prometheus/prometheus/pull/17071 is my attempt.

**Comment by bwplotka** [comment]
I still see 

```
// The type of samples (float64, histogram, etc) appended for a given series must remain same within an Appender.
// The behaviour is undefined if samples of different types are appended to the same series in a single Commit().
```

Is behaviour still undefined? I think it is defined by the fix: https://github.com/prometheus/prometheus/pull/17071

Can anyone give TL;DR on what's the expected behaviour now? Also is this NOT must now?

**Comment by beorn7** [comment]
I think the comment is outdated. @krajorama will create a PR (as per Slack communication).
